### PR TITLE
SONARJAVA-1901 Fix substitution in thrown types

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-frontend/src/test/java/org/sonar/java/resolve/SymbolTableTest.java [hunk 1]
```diff
@@ -1222,8 +1223,7 @@ public void infer_stream_types_on_chained_map() {
 
     IdentifierTree map = result.referenceTree(63, 8);
     MethodJavaType methodJavaType = (MethodJavaType) map.symbolType();
-    // expression type is correctly infered but method type is not recomputed and thus is still defered
-    assertThat(methodJavaType.resultType.isTagged(JavaType.DEFERRED)).isTrue();
+    assertThat(methodJavaType.resultType.isTagged(JavaType.DEFERRED)).isFalse();
```
Comment (benzonico) [inline java-frontend/src/test/java/org/sonar/java/resolve/SymbolTableTest.java]: Maybe we can also have a positive assertion ? testing what is expected not only what is not expected ?

File: java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java [hunk 2]
```diff
@@ -279,6 +279,8 @@ private void inferReturnTypeFromInferedArgs(MethodInvocationTree tree, Resolve.E
         MethodJavaType methodType = (MethodJavaType) resolution.type();
         if(!methodType.resultType.isTagged(JavaType.DEFERRED)) {
           registerType(tree, resolve.applySubstitution(methodType.resultType, typeSubstitution));
+          // change type associated to identifier as it may have been inferred in the process
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java]: "in the process" -> no idea what this means.

File: java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java [hunk 3]
```diff
@@ -279,6 +279,8 @@ private void inferReturnTypeFromInferedArgs(MethodInvocationTree tree, Resolve.E
...
+          registerType(identifier, resolution.type());
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java]: shouldn't  resolution.type() be replaced by methodType ?

Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java]: Am I correct in saying that you ended up associating methodType to the identifier of the method in order to access it from check ? That is far from obvious and deserve at least a comment (I am even wondering about a fixme here, this sounds a bit like a hack to shove the methodType into the semantic API)

Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java]: Ok, so this was already done before, but please update tests properly !

File: java-frontend/src/test/java/org/sonar/java/resolve/SymbolTableTest.java (hunk 1 above)
Comment (benzonico) [inline java-frontend/src/test/java/org/sonar/java/resolve/SymbolTableTest.java]: This is actually _really_ important  : this is the kind of uncomplete udpate that can lead to unfailing tests after some changes. This test is now lacking assertions.

File: java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java (hunk 3 above)
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java]: Correct. Was already done for method resolution. I'll update the tests.

File: java-frontend/src/test/java/org/sonar/java/resolve/SymbolTableTest.java (hunk 1 above)
Comment (m-g-sonar) [inline java-frontend/src/test/java/org/sonar/java/resolve/SymbolTableTest.java]: agreed. I took a shortcut here.

File: java-frontend/src/test/files/sym/ThrownTypeVariables.java [hunk 4]
```diff
@@ -1,7 +1,7 @@
 class A<T extends Throwable> {
   void foo() throws T {
-  }
 
+  }
```
Comment (benzonico) [inline java-frontend/src/test/files/sym/ThrownTypeVariables.java]: useless change.

File: java-frontend/src/test/files/sym/ThrownTypeVariables.java [hunk 5]
```diff
@@ -24,4 +24,3 @@ void test(MyInterface<java.util.NoSuchElementException> s) {
     new B<String>().bar(java.io.IOException::new);
   }
 }
-
```
Comment (benzonico) [inline java-frontend/src/test/files/sym/ThrownTypeVariables.java]: useless change

**Pull request merged**